tempesta_fw/sock.c

```diff
@@ -509,38 +533,59 @@ ss_tcp_process_data(struct sock *sk)
 			 * is returned back to the Linux kernel.
 			 */
 			if (r < 0) {
-				SS_DBG("Bad socket %p data processing, %d\n",
-				       sk, r);
+				SS_DBG("Processing error: sk %p r %d\n", sk, r);
 				goto out; /* connection dropped */
 			}
 			tp->copied_seq += count;
 			processed += count;
 
+			if (r == SS_STOP) {
+				if (tcp_fin) {
+					SS_DBG("Data FIN: sk %p\n", sk);
+					++tp->copied_seq;
+					goto out;
+				}
+				SS_DBG("Stop processing: sk %p\n", sk);
+				break;
+			}
+			if (frag_list) {
+				off = 0;
+				skb = frag_list;
+				frag_list = frag_list->next;
+				skb->next = NULL;
+				goto repeat;
```

Comment:
This is internal loop on skb fragments and you have 3 places where you do iterations over the list. The function also becomes to long, so it must be reworked: current loop over ```sk->sk_receive_queue``` stays in the function while the loop over ```skb->frag_list``` must be moved to other function and replaced by normal loop construction instead of goto. Probably the body of the internal loop is good to move to 3rd function.